### ISSUE TITLE
Process rabbit_queue_master_balancer with 0 neighbours exited with reason: no match of right hand value

### ISSUE THREAD

**Issue by petvaa01** [Ayanda-D/rabbitmq-queue-master-balancer#6]
I successfully tested balancing queue masters in small sandbox but when running on bigger environment I got the following error. Both environments have RabbitMQ 3.7.7 (Erlang 20.3.4) and running 5 node cluster.

```
rabbitmq-0 rabbitmq ** Last event in was '$balance_queues'
rabbitmq-0 rabbitmq ** When State == balancing_queues
rabbitmq-0 rabbitmq **      Data  == {state,<0.17117.399>,balancing_queues,0,37,58,7,10,10000,1000,1534151210278}
rabbitmq-0 rabbitmq ** Reason for termination =
rabbitmq-0 rabbitmq ** {{badmatch,{error,{noproc,{gen_server2,call,[<14607.1427.0>,sync_mirrors,infinity]}}}},[{rabbit_queue_master_balancer,balancing_queues,2,[{file,"src/rabbit_queue_master_balancer.erl"},{line,198}]},{gen_fsm,handle_msg,8,[{file,"gen_fsm.erl"},{line,483}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
rabbitmq-0 rabbitmq 2018-08-13 09:06:55.422 [error] <0.17157.399> CRASH REPORT Process rabbit_queue_master_balancer with 0 neighbours exited with reason: no match of right hand value {error,{noproc,{gen_server2,call,[<14607.1427.0>,sync_mirrors,infinity]}}} in rabbit_queue_master_balancer:balancing_queues/2 line 198 in gen_fsm:terminate/7 line 602
rabbitmq-0 rabbitmq 2018-08-13 09:06:55.422 [error] <0.17117.399> Supervisor rabbit_queue_master_balancer_sup had child rabbit_queue_master_balancer started with rabbit_queue_master_balancer:start_link() at <0.17157.399> exit with reason no match of right hand value {error,{noproc,{gen_server2,call,[<14607.1427.0>,sync_mirrors,infinity]}}} in rabbit_queue_master_balancer:balancing_queues/2 line 198 in context child_terminated
rabbitmq-0 rabbitmq 2018-08-13 09:08:23.809 [error] <0.32216.399> ** State machine rabbit_queue_master_balancer terminating
rabbitmq-0 rabbitmq ** Last event in was '$balance_queues'
rabbitmq-0 rabbitmq ** When State == balancing_queues
rabbitmq-0 rabbitmq **      Data  == {state,<0.17117.399>,balancing_queues,37,6,58,6,10,10000,1000,1534151298672}
rabbitmq-0 rabbitmq ** Reason for termination =
rabbitmq-0 rabbitmq ** {{badmatch,{error,{noproc,{gen_server2,call,[<14605.1528.0>,sync_mirrors,infinity]}}}},[{rabbit_queue_master_balancer,balancing_queues,2,[{file,"src/rabbit_queue_master_balancer.erl"},{line,198}]},{gen_fsm,handle_msg,8,[{file,"gen_fsm.erl"},{line,483}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
rabbitmq-0 rabbitmq 2018-08-13 09:08:23.809 [error] <0.32216.399> CRASH REPORT Process rabbit_queue_master_balancer with 0 neighbours exited with reason: no match of right hand value {error,{noproc,{gen_server2,call,[<14605.1528.0>,sync_mirrors,infinity]}}} in rabbit_queue_master_balancer:balancing_queues/2 line 198 in gen_fsm:terminate/7 line 602
rabbitmq-0 rabbitmq 2018-08-13 09:08:23.810 [error] <0.17117.399> Supervisor rabbit_queue_master_balancer_sup had child rabbit_queue_master_balancer started with rabbit_queue_master_balancer:start_link() at <0.32216.399> exit with reason no match of right hand value {error,{noproc,{gen_server2,call,[<14605.1528.0>,sync_mirrors,infinity]}}} in rabbit_queue_master_balancer:balancing_queues/2 line 198 in context child_terminated
```

**Comment by Ayanda-D** [comment]
Hi @petvaa01 can you give more detail on "bigger environment ". How many queues in particular, are active on the 5-node cluster?

**Comment by petvaa01** [comment]
Hi @Ayanda-D. Queue count is not that different between envs, 52 vs. 58, and set up similarly. What is different is that sandbox was basically idling with mostly empty queues and 100 connections, while other environment is in active use with messages flowing, queues with 0-120 messages (one queue has 24K messages) and about 350 connections. Thank you.

**Comment by bwmills** [comment]
Hi @Ayanda-D - testing the plugin in a non-production environment for evaluation.

### Note

Thank you for creating the plugin - the operational control it offers is exactly what we are looking for for our production environment. 

### Plugin

* Version: `0.0.3`
* Configs: see results below

### Environment

* Google Cloud Platform: `Kubernetes`
* Container type: `Docker`
* OS: `Ubuntu 16.04.3 LTS`

### RabbitMQ Config

* RabbitMQ version: `3.7.4`
* Erlang version: `20.3`
* Cluster size: `3 nodes`
* Number of queues: `134` (in non-production env)
* Typical queue depth: `range [0 - ~800,000]`
* Prefetch (server-side): `250`
* Mirroring: all queues mirrored
* Message counts: most (but not all) queues at `0` for this test
* HA Policies:
   - ha-mode: `exactly`
   - ha-params: `2`
   - ha-sync-batch-size: `20000`
   - ha-sync-mode: `automatic`
   - queue-master-locator: `min-masters`
   - queue-mode: `lazy`
* Plugins:
   -  rabbitmq_management
   - rabbitmq_management_agent
   - rabbitmq_web_dispatch
   - rabbitmq_queue_master_balancer
   - rabbitmq_auth_backend_http
* SSL: `enabled`

### Results 

Proceeded with steps per the ReadMe:
```
rabbitmqctl eval 'rabbit_queue_master_balancer:info().'
[{parent_pid,<9255.20874.1>},
 {phase,idle},
 {total,0},
 {position,undefined},
 {balanced,0},
 {operational_priority,5},
 {sync_timeout,3000},
 {policy_transition_delay,1000},
 {last_balance_timestamp,undefined}]
```
```
rabbitmqctl eval 'rabbit_queue_master_balancer:status().'
[{process_id,<9255.20875.1>},
 {queues_pending_balance,0},
 {memory_utilization,2912}]
```
```
rabbitmqctl eval 'rabbit_queue_master_balancer:load_queues().'
{ok,134}
```
```
rabbitmqctl eval 'rabbit_queue_master_balancer:go().'
```
At first run (after ~40 seconds):
```
E0000 2018-09-20T22:51:46.755Z 0000 rabbitmq] ** State machine rabbit_queue_master_balancer terminating
** Last event in was '$balance_queues'
** When State == balancing_queues
**      Data  == {state,<0.20874.1>,balancing_queues,69,37,134,14,5,3000,1000,1537483901649}
** Reason for termination =
** {{badmatch,{error,{noproc,{gen_server2,call,[<0.26510.1>,sync_mirrors,infinity]}}}},[{rabbit_queue_master_balancer,balancing_queues,2,[{file,"src/rabbit_queue_master_balancer.erl"},{line,198}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
E0000 2018-09-20T22:51:46.755Z 0000 rabbitmq] CRASH REPORT Process rabbit_queue_master_balancer with 0 neighbours exited with reason: no match of right hand value {error,{noproc,{gen_server2,call,[<0.26510.1>,sync_mirrors,infinity]}}} in rabbit_queue_master_balancer:balancing_queues/2 line 198 in gen_fsm:terminate/7
E0000 2018-09-20T22:51:46.756Z 0000 rabbitmq] Supervisor rabbit_queue_master_balancer_sup had child rabbit_queue_master_balancer started with rabbit_queue_master_balancer:start_link() at <0.20875.1> exit with reason no match of right hand value {error,{noproc,{gen_server2,call,[<0.26510.1>,sync_mirrors,infinity]}}} in rabbit_queue_master_balancer:balancing_queues/2 line 198 in context child_terminated
```
### Additional Notes

* The plugin is enabled on all three nodes but the commands were only executed on one node, after exec into the container

* While queues in our production env can carry a fair number of messages, for this test, the queues that contained messages had a messages_ready count of ~100

* The 3-node cluster was operating normally at the time of the test - i.e. no issues such as network partition, etc.

* There were no active consumers at the time of the test

**Comment by Ayanda-D** [comment]
@petvaa01 @bwmills thanks for reporting, and the valuable feedback. Made time to focus on this (sorry took a while). Now fixed in https://github.com/Ayanda-D/rabbitmq-queue-master-balancer/commit/3b76e04c9e6271ab719b83caf3fe985bdc2a3e9c and shipped in [v0.0.4 release](https://github.com/Ayanda-D/rabbitmq-queue-master-balancer/releases/tag/v0.0.4). Some elements may be extended and made more configurable, for increased operator flexibility. We'll add these based on demand and need. thanks!
